Allow specifying installation directory

krnowak opened the pull request:
### Requirements

* Filling out the template is required. Any pull request that does not include enough information to be reviewed in a timely manner may be closed at the maintainers' discretion.
* All new code requires tests to ensure against regressions

### Description of the Change

Currently the `--install` flag of the `./script/build` is a boolean flag that either installs atom to some predefined location or skips it. This PR extends the flag so it can accept a value telling the script where to install atom. Passing just `--install` will behave as it used to so far, so it will install atom to `/usr/local` on linux and other default locations on darwin and windows. Same with not passing the flag - atom won't be installed. But passing `--install=/opt/atom` will install atom under the given location.

The documentation is updated accordingly.

### Alternate Designs

I haven't considered other designs, extending the `--install` flag looked natural to me.

### Why Should This Be In Core?

It's a part of the builder script, so it can't be a package.

### Benefits

One can install the code in the location of choice. (Like for flatpaks, we want to install atom under `/app`).

### Possible Drawbacks

It's backward compatible, so not sure if there are drawbacks.

### Applicable Issues

Didn't file any.

krnowak committed:
Allow specifying installation directory

Ben3eeE commented:
https://github.com/atom/atom/issues/13240

ungb commented:
Tested on Mac, Linux, Windows.  Only issue I saw was if you use `~`, it doesn't work and doesn't do anything.  It even shows like it works on mac but doesn't.

see screenshot:
![image](https://cloud.githubusercontent.com/assets/5624870/24382137/2579f840-1309-11e7-8635-1ab0a90d90e9.png)


I tested both default `--install` and with a path.  Both worked on all.  On CentOs, regular `--install` didn't work due to permission issue, but I am seeing this on master as well
see:
![image](https://cloud.githubusercontent.com/assets/5624870/24382438/1b181a38-130b-11e7-9da2-58ad3c4bf79d.png)


Since this isn't related,  The only issue is using ~ as I think it's a fairly common scenario and seems to fail silently which isn't great.

krnowak committed:
Handle tilde as home dir in darwin and linux

krnowak commented:
Hey, @ungb, thanks for testing! I didn't thought about handling the tilde, probably because I thought that nodejs will handle this itself, but apparently it is not [nor is planning to](https://github.com/nodejs/node/issues/684).

I updated the PR - rebased it to recent master, replaced `!=` with `!==` and added some code for handling tilde (this adds a dep on the [`passwd-user`](https://www.npmjs.com/package/passwd-user) module). Note that the tilde expansion happens only on linux and mac, and I have tested it only on linux. I checked the following (`kv` is my user name):
- `~`
- `~kv`
- `~/`
- `~kv/`
- `~/usr`
- `~kv/usr`
- `~yaddayaddanosuchuser/blah`
  - `Error: Failed to expand the tilde in ~yaddayaddanosuchuser/blah - user "yaddayaddanosuchuser" does not exist`

thomasjo commented:
You can greatly simplify this by using the [untildify](https://github.com/sindresorhus/untildify) package.

j-f1 commented:
@thomasjo That only supports `~/...`, not `~user/...` AFAICT.

krnowak commented:
I considered the untildify package, but it didn't do the `~user` stuff.

Also, the appveyor failure seems unrelated, it even skips the installation, because `--install` flag was not passed.

ungb commented:
LGTM!  I was able to verify the new changes on linux and mac.

ungb commented:
Merging as I don't believe this is causing build failures on appveyor as I see other failures for similar things.
